### DISCUSSION TITLE
Cannot connect to Trino via OAUTH2 with SSO credentials

### DISCUSSION THREAD

**Discussion by lejanco** [trinodb/trino-python-client#437] [Q&A]
I am trying to connect to trino via oauth2 using the externalauthentication instructions in the README(e.g. keyring). I have tried (what seems like) everything. You can see the commented out efforts here as well.  I have done the pip install 'trino[<keyring>]'

` trino_connection = trino.dbapi.connect(host=host,
                                               port=port,
                                               user=username,
                                               catalog=catalog,
                                               schema=schema,
                                               auth=trino.auth.OAuth2Authentication(
                                                   # redirect_auth_url_handler=redirect_auth_url_handler
                                               ),
                                               http_scheme=http_scheme,
                                               # verify=False,
                                               # extra_credential=[('user', username),
                                               #                   ('password', password),
                                               #                   ("SSL", "true"),
                                               #                   ("SSLVerification", "NONE"),
                                               #                   ("externalAuthentication", "true")
                                               #                   ]
                                               )`

The debug suggests the token is being retrieved but not being authenticated. Help is greatly appreciated, but please limit suggestions that require admin rights to trino as I do not have that authorization. Thanks:

DEBUG:

**Comment by lejanco** [comment]
I guess debug didn't paste. Here is a sample (redacted)
trino_df = pd.read_sql_query(sql=read_trino_sql, con=sql_alchemy_engine)
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1):<host>:443
DEBUG:urllib3.connectionpool:https://<host:port> "POST /v1/statement HTTP/1.1" 401 12
DEBUG:urllib3.connectionpool:https://<host:port>  "GET /oauth2/token/<token number> HTTP/1.1" 200 127
DEBUG:urllib3.connectionpool:https://<host:port>  "GET /oauth2/token/<token number> HTTP/1.1" 200 127

**Comment by mosabua** [comment]
@lejanco can you please fix the code snippets in the descriptions so it is actually readable and we can compare it to the docs in the readme at https://github.com/trinodb/trino-python-client?tab=readme-ov-file#oauth2-authentication

**Reply by lejanco** [reply]
@mosabua, please see the code snippet below:

So far I have tried:

- Successfully installed keyring: pip install trino[external-authentication-token-cache].  
- included/excluded the redirect_auth_url_handler 
- Used a dictionary format for the extra_credential arguement instead of a list of tuples
- included/excluded verify=False argument

> 
      trino_connection = trino.dbapi.connect(
                                     host=host,
                                     port=port, 
                                     user=username, 
                                     catalog=catalog, 
                                     schema=schema,
                                     verify=False,
                                     http_scheme=http_scheme,
                                     auth=trino.auth.OAuth2Authentication( 
                                      # redirect_auth_url_handler=redirect_auth_url_handler 
                                                           ),
                                     # extra_credential=[('user', username), 
                                     #                   ('password', password), 
                                     #                   ("SSL", "true"), 
                                     #                   ("SSLVerification", "NONE"), 
                                     #                   ("externalAuthentication", "true") 
                                     #                   ] 
                                                           )
> 

The debug code (snippet) is the same:
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1)::443
DEBUG:urllib3.connectionpool:https://host:port "POST /v1/statement HTTP/1.1" 401 12
DEBUG:urllib3.connectionpool:https://host:port "GET /oauth2/token/ HTTP/1.1" 200 127
DEBUG:urllib3.connectionpool:https://host:port "GET /oauth2/token/ HTTP/1.1" 200 127....

**Reply by hashhar** [reply]
Since the authentication happens on the server check the server logs to see what error is being reported there. Note that for authentication to work on the server it needs to be configured using HTTPS, Trino won't accept passwords over HTTP.

Also what error do you get in the Python script?

In general the correct invocation is:

```python
conn = connect(
    host=host,
    port=port,
    user=username,
    catalog=catalog,
    schema=schema,
    auth=OAuth2Authentication()
)
```

You only need to add `verify=False` if the Trino server is using self-signed TLS certificates. And `OAuth2Authentication()` doesn't need any parameters.

**Reply by lejanco** [reply]
@hashhar

How would I check the server logs? Just fyi, i don't have admin access if that's a pre-requisite. 

For authentication reference, in order to connect via Oauth in dbeaver, we use the following parameters/arguments: 

**Driver Properties:**
- SSL: true
-  externalAuthentication: true

**User Properties**
- SSLVerification: NONE


**The full traceback:**

DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): <host>:<port>
DEBUG:urllib3.connectionpool:https://<host>:<port> "POST /v1/statement HTTP/1.1" 401 12
DEBUG:urllib3.connectionpool:https://<host>:<port> "GET /oauth2/token/<token#> HTTP/1.1" 200 104
DEBUG:urllib3.connectionpool:https://<host>:<port> "GET /oauth2/token/<token#> HTTP/1.1" 200 104
DEBUG:urllib3.connectionpool:https://<host>:<port> "GET /oauth2/token/<token#> HTTP/1.1" 200 104
DEBUG:urllib3.connectionpool:https://<host>:<port> "GET /oauth2/token/<token#> HTTP/1.1" 200 104
DEBUG:urllib3.connectionpool:https://<host>:<port> "GET /oauth2/token/<token#> HTTP/1.1" 200 104
INFO:trino.client:failed after 1 attempts
Traceback (most recent call last):
  File "C:\FOLDER\DEV\TMP\ds\.virtualenv\reporting_server_modernization\lib\site-packages\pandas\io\sql.py", line 2202, in execute
    cur.execute(sql, *args)
  File "C:\FOLDER\DEV\TMP\ds\.virtualenv\reporting_server_modernization\lib\site-packages\trino\dbapi.py", line 595, in execute
    self._iterator = iter(self._query.execute())
  File "C:\JFOLDER\DEV\TMP\ds\.virtualenv\reporting_server_modernization\lib\site-packages\trino\client.py", line 806, in execute
    response = self._request.post(self._query, additional_http_headers)
  File "C:\FOLDER\DEV\TMP\ds\.virtualenv\reporting_server_modernization\lib\site-packages\trino\client.py", line 555, in post
    http_response = self._post(
  File "C:\FOLDER\DEV\TMP\ds\.virtualenv\reporting_server_modernization\lib\site-packages\trino\client.py", line 899, in decorated
    raise error
  File "C:\FOLDER\DEV\TMP\ds\.virtualenv\reporting_server_modernization\lib\site-packages\trino\client.py", line 886, in decorated
    result = func(*args, **kwargs)
  File "C:\FOLDER\DEV\TMP\ds\.virtualenv\reporting_server_modernization\lib\site-packages\requests\sessions.py", line 637, in post
    return self.request("POST", url, data=data, json=json, **kwargs)
  File "C:\FOLDER\DEV\TMP\ds\.virtualenv\reporting_server_modernization\lib\site-packages\requests\sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
  File "C:\FOLDER\DEV\TMP\ds\.virtualenv\reporting_server_modernization\lib\site-packages\requests\sessions.py", line 710, in send
    r = dispatch_hook("response", hooks, r, **kwargs)
  File "C:\FOLDER\DEV\TMP\ds\.virtualenv\reporting_server_modernization\lib\site-packages\requests\hooks.py", line 30, in dispatch_hook
    _hook_data = hook(hook_data, **kwargs)
  File "C:\FOLDER\DEV\TMP\ds\.virtualenv\reporting_server_modernization\lib\site-packages\trino\auth.py", line 302, in _authenticate
    self._attempt_oauth(response, **kwargs)
  File "C:\FOLDER\DEV\TMP\ds\.virtualenv\reporting_server_modernization\lib\site-packages\trino\auth.py", line 339, in _attempt_oauth
    token = self._get_token(token_server, response, **kwargs)
  File "C:\FOLDER\DEV\TMP\ds\.virtualenv\reporting_server_modernization\lib\site-packages\trino\auth.py", line 382, in _get_token
    raise exceptions.TrinoAuthError("Exceeded max attempts while getting the token")
trino.exceptions.TrinoAuthError: Exceeded max attempts while getting the token
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "C:\FOLDER\DEV\TMP\ds\.virtualenv\reporting_server_modernization\lib\site-packages\pandas\io\sql.py", line 2206, in execute
    self.con.rollback()
  File "C:\FOLDER\DEV\TMP\ds\.virtualenv\reporting_server_modernization\lib\site-packages\trino\dbapi.py", line 242, in rollback
    raise RuntimeError("no transaction was started")
RuntimeError: no transaction was started
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "C:\FOLDER\DEV\TMP\ds\.virtualenv\reporting_server_modernization\lib\site-packages\IPython\core\interactiveshell.py", line 3508, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-6-042fc92444ed>", line 259, in <module>
    trino_df = pd.read_sql_query(sql=read_trino_sql, con=sql_alchemy_engine, )
  File "C:\FOLDER\DEV\TMP\ds\.virtualenv\reporting_server_modernization\lib\site-packages\pandas\io\sql.py", line 469, in read_sql_query
    return pandas_sql.read_query(
  File "C:\FOLDER\DEV\TMP\ds\.virtualenv\reporting_server_modernization\lib\site-packages\pandas\io\sql.py", line 2266, in read_query
    cursor = self.execute(sql, params)
  File "C:\FOLDER\DEV\TMP\ds\.virtualenv\reporting_server_modernization\lib\site-packages\pandas\io\sql.py", line 2211, in execute
    raise ex from inner_exc
pandas.errors.DatabaseError: Execution failed on sql: 
                        select * 
                        from <catalog>.<schema>.<table>
                         
Exceeded max attempts while getting the token
unable to rollback

**Reply by hashhar** [reply]
> How would I check the server logs? Just fyi, i don't have admin access if that's a pre-requisite.

Ask whoever is managing your Trino server to get them for you.

The error from the client is:

```
trino.exceptions.TrinoAuthError: Exceeded max attempts while getting the token
```

So are you running the Python script on a server or a local machine? Does it open a browser window for you when you run the script to perform the OAuth login?
OAuth by definition requires interacting with server to finish the login process so if the failure happenned on the server (looks like it from the logs) you need the server logs to find it's cause.

**Reply by hashhar** [reply]
Also according to the DBeaver configs you mentioned you would need something like:

```
trino_connection = trino.dbapi.connect(
			 host=host,
			 port=port, 
			 user=username, 
			 catalog=catalog, 
			 schema=schema,
			 verify=False, # SSLVerification=NONE
			 http_scheme="https", # SSL=True
			 auth=trino.auth.OAuth2Authentication() # implies externalAuthentication=True
		)
```

**Reply by lejanco** [reply]
@hashhar 
To Answer your questions
So are you running the Python script on a server or a local machine?

- Currently running of local

 Does it open a browser window for you when you run the script to perform the OAuth login?

- Only if I specify it to open: webbrowser.open_new(r"https://<host>/oauth2/callback/")

I did try the script above originally and the traceback is the same.

**Reply by hashhar** [reply]
> Only if I specify it to open: webbrowser.open_new(r"https:///oauth2/callback/")

This is unexpected. By default if you use `auth=trino.auth.OAuth2Authentication()` it should open a web browser window to perform the OAuth login, if the browser cannot be opened then it prints a URL which you need to manually open. From the traceback you shared it doesn't seem like either of it happened.

Without the server logs it's almost impossible to tell what is going wrong since it's the server who receives the token and responds to the client.

**Reply by matialm** [reply]
@hashhar the redirect callback it's not working because when try to get x_redirect_server it's returning None because the parser return "Bearer x_redirect_server" instead "x_redirect_server". I trying change that to test it and it works.

The problem is in auht.py file _attempt_oauth method.

I will try to create a pr to fix it.

**Reply by matialm** [reply]
This should fix the problem
https://github.com/trinodb/trino-python-client/pull/445
